### DISCUSSION TITLE
What About a type-safe SDK?

### DISCUSSION THREAD

**Discussion by a4amado** [pocketbase/pocketbase#6138] [Q&A] [answered]
I know this is a hobby project, and even if it becomes considered, it would probably be at least a couple of years in the future. But it would be great if it were possible.

**Comment by chnfyi** [comment, marked as answer]
You can define a global PocketBase type using type assertion.

```typescript
interface Task {
  id:   string;
  name: string;
}

interface Post {
  id:     string;
  title:  string;
  active: boolean;
}

interface TypedPocketBase extends PocketBase {
  collection(idOrName: string): RecordService // default fallback for any other collection
  collection(idOrName: 'tasks'): RecordService<Task>
  collection(idOrName: 'posts'): RecordService<Post>
}

...

const pb = new PocketBase("http://127.0.0.1:8090") as TypedPocketBase;

pb.collection('tasks').getOne("RECORD_ID") // -> results in Promise<Task>
pb.collection('posts').getOne("RECORD_ID") // -> results in Promise<Post>
``` 

As an alternative you can take a look to some initiative like [pocketbase-typegen](https://github.com/patmood/pocketbase-typegen).

**Reply by tigawanna** [reply]
This works great for me

V < 0.23
https://github.com/david-plugge/typed-pocketbase

V 0.23   https://github.com/tigawanna/typed-pocketbase

**Comment by ouvreboite** [comment]
Sorry for reviving this, but given that several "type" generators already exist (so the need is there), but are of various quality and often out-of-date with the main project (do any of them support the new geopoint type?), maybe it would make sense to have some code in pocketbase itself to generate the Typescript types/helper automatically?

Having the typegen in pocketbase would also allow to have a better sync with change made to the DB. For example it could run at the same time the migration files as generated.

There are two main things to type 
- the records, to have strongly typed response (and input)
  - including expanded relation would be nice, at least for 1st level relations and back relations
  - this seems mostly a solved problem, so not too much complexity here
- the `sort`, `filter` and `expand` parameters.
  - that can be a bit tricky, as the current type is simply `string`
  - `filter` seems especially complex

One way to provide better typing for `sort`, `filter` and `expand` without being too intrusive could be to generate typed-ish builders.
  
Example 

```typescript
//example of autogenerated helper for a "post" collection
//additional generic code is hidden
export const Post = {
    Options: (
        { sort, expand, filter, ... others}: Options<
            "+created" | "-created" | "+author" | "-author", //sort
            "author" | "comments_via_post"> //expand
    ): RecordFullListOptions => {
        const result: RecordFullListOptions = {...others};
        if (sort?.length) result.sort = sort.join(',');
        if (expand?.length) result.expand = expand.join(',');
        if (filter) result.filter = filter;
        return result;
    },
    Fields: {
        Id: TextField("id"),
        Title: TextField("title"),
        Status: SelectField<"draft" | "published" | "retracted">("status"),
        Author: ReferenceField("author", {
            Id: TextField("author.id"),
            Name: TextField("author.name"),
            Role: SelectField<"admin" | "reader" | "anon">("author.role"),
            //nested relations not included 
        }),
        //other fields
    }
}

//in use
pb.collection("post").getFullList(Post.Options({
    sort: ['+author', '+created'], //✅ the sort columns are an array of enum, so good autocompletion
    expand: ['author'], //✅ same for expand
    filter: And( //✅filter is a bit more complex, but still provide good typing
        Post.Fields.Author.Name(':lower', "=", "John Doe'"), //✅relation subfield, including support for modifiers
        Post.Fields.Status('=', "retracted"), //✅select fields are backed by an enum
        "created > @yearStart" //🔓 can opt out and include custom strings
    ),
    skipTotal: true
}));

```

Post.Options() would simply generate the correct strings for sort/filter/expand

```javascript
{
  "sort": "+author,-created",
  "expand": "author",
  "filter": "(author.name:lower = 'John Doe' && status = 'retracted' && created > @yearStart)"
  "skipTotal": true,
} 
```

In action:

https://github.com/user-attachments/assets/7300b63f-e812-4577-bcf9-77352bb8cb12

**Reply by ouvreboite** [reply]
Typescript playground of what that could look like: [link](https://www.typescriptlang.org/play/?#code/PTBmEMGsFMAIAcD2BjGAXARuAznNBPeOURAJ1jQAs4ATaAW0QCgCjYAlaZMmgMQFcANoIAyAS2xoA8vDRjEAO2ywAvLADeTWNoTgA5tAD8ALlgL+9DNFIBuLTqKkACvqOnzl63Z2xsZNCa+aKRiCnreOtAAHvDgCjSBkiFhEdqgYoJo1onBoeH22tiQYvAAKoho4IKBGIiIgtBxqbBYaMiUgR5WtkwAvkxM3EpoCBiqGgWw3MJccoqmABTTDchzCqZJeQCUqgB8GrCk0Gj8pAoTPj4GaALC4pKLiLLySqac3KR8QqIS0s+K2B2Kn26n6Pl6-V6sBw0IU+DsTBA3HojHOgkQejEyBYhDgAFkhHJ4A0ZNZwGgyOMAEQAPypsAAPrAqQBCOmM5mGOl2VhwACiAEd+FUxARSaRyZS1FSVPSmazZTzcbAAMKIeixEJ+BTiyXkaW7OXM3ayjlUgA8RotipxbAAchZurqKfrYILhYJRfhnZSmWqNeAtYofT1ebAQ+N3SKxY49Rz-ZqJMHYy6OQTMiUSSmyDZYCAAHSFgagfgKVYvWAAITqgl4YmgghoC3SDZoG1yYS2mh8Q0ksFA5zUCwWhfzgb02FMcXwAG0ALpAkGTHxiUCwBbj7D5hphKiqFRqAAMW2Xl20RxOZ379cbzRXa43pAn2+gu8o+7UACYdt2z5deyMM5PGSLoADSwAAblU-DQHO4ybtCygzlGnoxiBZDgbU9SNAoc53n+F6nOcAAGAAk6gto2ULkcBEoutR6hQYIMG9MR+HaGCf5UKQiAAO5mNA-F8qQPGkAsVIAKr0HErg0FSJ7glszSEVeA6IRcf4LFs7bJHooGntoCy0XqpgoV6IbgUxMGmFhDRxNpQS6fp4J2L0CIlmWaywKU0Q3DeTaUW2jnbL+2gAf2g7riOhablOcLzouGlnqu66bi+b4frAx4GZcKnnIF7GwClj7Pjueh7geX4-jlPYAoBxlgZB0GwfBT7KDCM4WcFYR4TVOh5bAZEUf5DENWQUIAOTkVZ0C9BNbE5VCDa4EVD5pWVFUHrAADM1V-v+dWwDOAD64FjaQlnNXBagIR1VLGOivHWFS4FdZsPWFT4A1DYFvQPXx1iwDR2akAxM2sYVnFntxfECUJIlkOJUkyQYckKTovRKZMA1qTCoWXFpOl5M5mnnaYXUzUTnZU3pOULIwNCrvWpCmPdj3PWdIPkyDl3MdANMOe9tMuX07mluWiiwA6nikHWrbNv5Av4+FalDtFY5tXFs4LnsSWXMV62vuV76VVl6P7eexxEderaFQbbXpcbmXfnr+3hUBPNNXz13Qm16mdZ7XTWL1Fv9VbV4-SNQPqOdYPNRDi05TD-EKIJboI2JknSQosnyZMmPKeH+XnHjdOCx2wuk1z4ae5TZiOtY5dOfnrli55FYAMoNrMcuNuapSwL5r40MoQu7ArrYC3rKuRcOo6xbC2uJfj96pQ7G0m1t2Wh5bl75f5dtrevRubVVrt-u7528zBPu3Uhb0VyHO-feRv3R7HsBTYx8fzZDSeUDxFOadhKiSRtnXO5tYAF2xkXCK6kV46EJt1SuZ4jLVzMmhOiGEvbWW8k3YmdMyawHTESLM6ELo4P5sg-BYQSYcVcsWcWXlOCgGsK+ZA0Be40H7uPQKNNwJHFAKYUoXZJgz3GHPGKmtF4JV1gg7Q9tSon03keSB+0cYHxyoorcG9nZ7R3pfWuV1WoTn9g-XST9Q4v2Gq2UaIM458wTvtKGlxk5w3TqArOKNoBo3zljHwUgMAACtZhjmwNgMQegFDNgUAI6AoB-Fhz3nA0u+0kFCzoQTIhGDvRGL5gLfh+dYAADJvKt0Ye3SWABxaAiAnCIFCH5eWfDqF6wGvI2AghySmGlt0LhE9GywAANTMm3OSeSmSfDonWFLBust-IDJoMM0Z0yJn51FgMcKdZMjWGUGofGABBeIixRzpB2aQScyCEo011sRBYr8MhZAufmIJDTokTRKaUiaWxMbEUyVIFm65TmPN2TTa5yDbn3OGucrcrzQgLA+QyJk3zfmZLtBURYZynkFIhcCQaLIoVYusKiyYwAABUZLYAAB0AACRwhTQD7GS4AkxSL0pgg8c+2gDn8CoG8eJbCyycIWVSOl0AGWSDHLyygL0uWXAAJJBR8lEJpjZxJiolWgKVVB8xiDRpMv8dpwD0Cocq1VTZRXssZVq8A0r8w5xNWsneeZgCp0kD4w4DZyQvGUAoCoRUyzMToEsxakDeiZJoOSaAeJwDIB4pcjp5LKW0vwI0UgHdKikBGMynKpFU2BgzYGNArMaX5vTZmtAL0cpJupaWtNfJ4iwBzftPN9bjnMjrYGBtclMnhr6AMMMMg1jYH7h3fwg8VXD1HhXcCpQ+QxDiEsoe8Rp26X2PsyYfgs2BFKGOrN85mjRFiPEHd87j00APZMIlpAci6QRD4EAsRwkw34HoSgkxYgGE6HM5ojgXBfvcD+zdxQygVCqDUGsOFmitHaN+mW5TETAFtRSAw+VKRID7MsWYLxBiHXqX2DdAT-ivHXDldQvh-DgSPYu8C17wKjgqNQcgf0a7DvNH1bQVIhlxsaFkOSZoAC0PGo38flEM5DlAyBWgExJqT4EQBbrQBx5ksnSBWmRCahQaBsDHQgmIcAx0MOVv2CAaj8RJgOXeDwW4PxJBDp9XInK4UjjYEJPyj4Xw7i-HswCcY5GGNUEBm5LRD5FOGEdlQHYLnCT5kU+MRTLy3kItAt8w+64zMJAi5QKLjKYsZfGBlxL8KJopcSclB816cuucyPma94xr2FV7NhbcGIFjRcyGV3KsD2toGaH2nwXCE05UVcI3y-SqR6qdX+UoooGijZVeNuQaAGhTbPIWk4lyu4rHNZaGgEpQCVrNPAfgGBPTYGoKJ5kF4JSrB8VSceVJJDkn4NgVblwVRHBE-N814lhN8bez4CS8BI18e++N-gwORMA50DyqgZB+WsKOEK8bqnZUdJ8CN7yY2RWqd1fq5TRqTVg5x9Ksg9rjXQGh2edg2FTBbZ7v5S04AaD0FCFaT7dA1NmjiIoe74lcc8RW1sA1Z4RAoG9fMWANS6lvJ+1SdEyAJcKCp5cEAbq+Oeq6cOsw-rQjICDR60NIuXWMcBoFbA6z+iIdCLAF70BcPDAOEcmgr0KHorQOBNl4qOUe9gCD6Nsb41QPGNsp52AETwAwPmLDEtolUiM-JfM1wbP3DQAsfDWqfNKAWPjRTpgZwTXE6T0gJXP5Cc+3xiac5MkZfzxNVTVfMnXtMM7hYOUM-5kG-mdbL2EUqFL1SPb4ADsq47132HknM6yk91ayVE+tLC-b4gSVXegf+6bBNXYpex-+S3B93jPiF+ZMX5jJSQA)

(scroll to the bottom to use Post.Options())
